### DISCUSSION TITLE
login and server action NextJS offline

### DISCUSSION THREAD

**Discussion by Clems02** [serwist/serwist#191] [Q&A]
Hello, I just started using this library to create a pwa using nextjs app router.
I have several questions that I can't find the answer to in the documentation (I admit I don't really understand the docs).
1. I saw that there is a second pwa library that you maintain. Which one is best suited to use with nextjs app router? (https://github.com/DuCanhGH/next-pwa)
2. When the user is offline, is it possible to keep their authentication active? For now, when they are offline, they have to go back to the login page, log in and refresh the page to find their "connected" status. I would like their status to remain connected without having to go back to the login page.
3. Is it possible to cache a server action when the user is not connected? I would like to couple this with a useOptimist hook to update the UI and then refresh the database once it gets internet.
4. Is there a way to test the offline mode of an application without having to put it in production?

Thanks for your help!

**Comment by DuCanhGH** [comment]
@Clems02 

> I saw that there is a second pwa library that you maintain. Which one is best suited to use with nextjs app router? (https://github.com/DuCanhGH/next-pwa)

Literally the first thing in README.md:

![image](https://github.com/user-attachments/assets/056d969f-133f-4b14-aa78-7c663323f949)

:D

> When the user is offline, is it possible to keep their authentication active? For now, when they are offline, they have to go back to the login page, log in and refresh the page to find their "connected" status. I would like their status to remain connected without having to go back to the login page.

Supposed that your user is trying to access a resource protected by your middleware, you only need to have a successful for each of those resources cached. Well, to ensure that nothing weird related to authorization happens, I suggest clearing the caches whenever your user logs out :)

If your app is fetching user info from the server and sending it to the client by inlining the info somewhere in the HTML responses, that info will be cached alongside said responses. Well, it's a part of those responses, after all.

If you're fetching user info on the client via a, let's say, "/user_info" endpoint, you can just cache the responses of that endpoint :) I suggest setting the `Vary` header so that it includes the header with which you use to identify the user, such as the `Cookie` or the `Authorization` header.

> Is it possible to cache a server action when the user is not connected? I would like to couple this with a useOptimist hook to update the UI and then refresh the database once it gets internet.

You can use the Background Synchronization API ([here is how to use it with Serwist](https://serwist.pages.dev/docs/serwist/guide/background-syncing)) to defer tasks to when your user is back online along with `Route.setCatchHandler` to provide a response that indicates that your user is offline, in which case you may optimistically update your UI then refresh it when your user is back online. But refrain from caching POST requests such as those which your application makes when using server actions.

As an example, this is how you may do it:

```ts
import { defaultCache } from "@serwist/next/worker";
import {
  Serwist,
  Route,
  NetworkOnly,
  BackgroundSyncPlugin,
} from "serwist";

const backgroundSync = new BackgroundSyncPlugin("myQueueName", {
  maxRetentionTime: 24 * 60, // Retry for a maximum of 24 hours (specified in minutes)
});

const route = new Route(
  ({ sameOrigin, url }) => sameOrigin && url.pathname === "/post",
  new NetworkOnly({
    plugins: [backgroundSync],
  }),
  "POST",
);

route.setCatchHandler(async () => Response.json({ type: "offline", message: "User is offline." }, { status: 200 }));

const serwist = new Serwist({
  // Other options...
  runtimeCaching: undefined,
});

// In the future, we will support passing a `Route` object to `runtimeCaching`, but this will do for now.
serwist.registerRoute(route);

for (const route of defaultCache) {
  serwist.registerCapture(route.matcher, route.handler, route.method);
}

serwist.addEventListeners();
```

> Is there a way to test the offline mode of an application without having to put it in production?

https://developer.chrome.com/docs/workbox/improving-development-experience

As for `@serwist/next` or `@serwist/vite`, you will need to copy the entirety of `defaultCache` to your service worker source for now. This is because in development, those packages set `defaultCache` to an array containing just a network-only route, which means anything offline won't work.

**Reply by Clems02** [reply]
@DuCanhGH 

Thank you very much for your answers. It seems really complex for my level I think...

> You can use the Background Synchronization API ([here is how to use it with Serwist](https://serwist.pages.dev/docs/serwist/guide/background-syncing)) to defer tasks to when your user is back online along with `Route.setCatchHandler` to provide a response that indicates that your user is offline, in which case you may optimistically update your UI then refresh it when your user is back online. But refrain from caching POST requests such as those which your application makes when using server actions.
> 

Are you saying that I have to create an API route instead of using a server action for this?

> As for `@serwist/next` or `@serwist/vite`, you will need to copy the entirety of `defaultCache` to your service worker source for now. This is because in development, those packages set `defaultCache` to an array containing just a network-only route, which means anything offline won't work.


For this I understood that it was necessary to copy all the defautcache code from next. I created a second file and copied everyone. I removed the first part of course in order not to take into account the development environment in order to have identical behavior to production. However I don't have the impression that this has changed anything.
I am currently trying to delete the cache when I modify my manifest file. I updated the name of the app for example and when I redeploy, I stay on the old manifest file (while in dev, the name has changed...).
I tried different chatgpt suggestion but nothing works, the file cache remains active. The only way found for the moment is to delete the cache manually from the browser settings ^^'

This is my custom cache strategy in my "localCache" file (copy to the defaultCache from serwist/next). But it's not working and the old manifest is still present.
```js
{
    matcher: /\.(?:json|xml|csv)$/i,
    handler: new NetworkFirst({
      cacheName: "static-data-assets",
      plugins: [
        new ExpirationPlugin({
          maxAgeSeconds: 1,
          maxAgeFrom: "last-used",
        }),
      ],
    }),
  },
```

and my sw.ts
```js
import type { PrecacheEntry, SerwistGlobalConfig } from "serwist";
import { Serwist } from "serwist";
import { localCache } from "./localCache";

declare global {
  interface WorkerGlobalScope extends SerwistGlobalConfig {
    __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: localCache,
});

serwist.addEventListeners();
```

I also saw in the doc that it was necessary to "delete" the old worker services in order to set up the new ones. Is this managed automatically by the lib or is it necessary to do something?

**Reply by DuCanhGH** [reply]
@Clems02 

> Are you saying that I have to create an API route instead of using a server action for this?

...No? A server action fundamentally is just a POST request to the same route you're on when you submit a form.

But if you are targeting all POST requests, the same code also works.

> I am currently trying to delete the cache when I modify my manifest file.

I suggest adding your manifest.json to `precacheEntries`. Whenever you update your service worker, your web app manifest will be updated along with it. If this file is located in the public directory, and you don't have `additionalPrecacheEntries` set, you don't have to do anything; otherwise, add the file to `additionalPrecacheEntries`.

> I also saw in the doc that it was necessary to "delete" the old worker services in order to set up the new ones. Is this managed automatically by the lib or is it necessary to do something?

It is done by the browser...

**Comment by Clems02** [comment]
> I suggest adding your manifest.json to precacheEntries. Whenever you update your service worker, your web app manifest will be updated along with it. If this file is located in the public directory, and you don't have additionalPrecacheEntries set, you don't have to do anything; otherwise, add the file to additionalPrecacheEntries.s`.

I'm really sorry but I can't do anything. I tried to put my manifest file in the additionalCacheEntries but nothing.
I also tried that, doesn't work either. I constantly stay on the old cache. I have to manually delete my browser cache to have the new version of my manifest.

```js
const CACHE_NAME = "cache-0.31";
const files = ["/manifest.json"];

self.addEventListener("install", function (e) {
  e.waitUntil(
    caches.open(CACHE_NAME).then(function (cache) {
      return cache.addAll(files);
    }),
  );
});

self.addEventListener("activate", function (event) {
  event.waitUntil(
    caches.keys().then(function (cacheNames) {
      return Promise.all(
        cacheNames
          .filter(function (name) {
            return name !== CACHE_NAME;
          })
          .map(function (cacheName) {
            return caches.delete(cacheName);
          }),
      );
    }),
  );
});
```
```js
import withSerwistInit from "@serwist/next";

/** @type {import('next').NextConfig} */
const nextConfig = {
  transpilePackages: ["lucide-react"], // add this
};

const withSerwist = withSerwistInit({
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  additionalPrecacheEntries: ["/manifest.json"],
});

export default withSerwist(nextConfig);
```

**Reply by DuCanhGH** [reply]
@Clems02 

> additionalPrecacheEntries: ["/manifest.json"],

A revision please. You can use the file's content hash. Without the revision, Serwist won't know if the /manifest.json in the cache is up-to-date or not. It should've warned you about that.

**Reply by DuCanhGH** [reply]
@Clems02 

```js
// @ts-check
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import withSerwistInit from "@serwist/next";
import globSync from "glob";

// Scan for files to cache in the public directory.
// A temporary workaround before v10.
const publicScan = globSync(["**/*"], {
  nodir: true,
  cwd: "public",
  ignore: ["swe-worker-*.js", "sw.js", "sw.js.map"],
});

/**
 * @param {string} filePath
 * @returns
 */
const getFileHash = (filePath) => crypto.createHash("md5").update(fs.readFileSync(filePath)).digest("hex");

const withSerwist = withSerwistInit({
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  additionalPrecacheEntries: [
    { url: "/manifest.json", revision: getFileHash("app/manifest.json") },
    ...publicScan.map((f) => ({
      // Replace "/" with your `nextConfig.basePath` if applicable.
      url: path.posix.join("/", f),
      revision: getFileHash(path.join("public", f)),
    })),
  ],
});

/** @type {import("next").NextConfig} */
const nextConfig = {
  reactStrictMode: true,
};

export default withSerwist(nextConfig);
```

**Reply by Clems02** [reply]
@DuCanhGH 

I just try that:

```js
import withSerwistInit from "@serwist/next";

/** @type {import('next').NextConfig} */
const nextConfig = {
  transpilePackages: ["lucide-react"],
};

const withSerwist = withSerwistInit({
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  additionalPrecacheEntries: [
    { url: "/manifest.json", revision: "manifestrevision" },
  ],
});

export default withSerwist(nextConfig);
```

and for the moment it's working ! Thank you a lot for your help. 

I really still need to learn to master the thing. I will probably still need a lot of help ^^'

I really have trouble with the documentation. Where is the information for review? In the page https://serwist.pages.dev/docs/next/configuring/additional-precache-entries , there is only the url mentioned.

Since I use nextJS app router, is there any point in looking at the basic serwist documentation?

https://serwist.pages.dev/docs/serwist
or can I keep only the serwist/nextJS document?

**Reply by DuCanhGH** [reply]
@Clems02 

> revision: "manifestrevision"

If the revision doesn't represent the actual file, you will be as screwed as if you didn't specify `revision` in the first place :)

> Where is the information for review?

The docs is still unfinished, so some information like this is not there yet. Moving forward to 9.1.0, we will renovate the docs in many places. Not now though. I'm still way too busy for that.

> or can I keep only the serwist/nextJS document?

Please read the documentation of `serwist` as well. It helps you a lot with questions like the ones you made in this discussion:

- https://serwist.pages.dev/docs/serwist/guide/precaching
- https://serwist.pages.dev/docs/serwist/runtime-caching
- https://serwist.pages.dev/docs/serwist/runtime-caching/caching-strategies
- https://serwist.pages.dev/docs/serwist/runtime-caching/routing
- https://serwist.pages.dev/docs/serwist/runtime-caching/plugins
- https://serwist.pages.dev/docs/serwist/guide/background-syncing
